tests/test_pythonpackage.py

```diff
@@ -0,0 +1,133 @@
+
+import os
+import shutil
+import sys
+import subprocess
+import tempfile
+
+from pythonforandroid.pythonpackage import (
+    get_package_dependencies,
+    get_dep_names_of_package,
+    get_package_name,
+    _get_system_python_executable,
+)
+
+
+def test_get_package_name():
+    assert get_package_name("PiLLOw") == "Pillow"
+
+
+def test_get_package_dependencies():
+    # Check that jinja2 is returned as direct dep:
+    assert(len([dep for dep in get_package_dependencies("python-for-android")
+                if "jinja2" in dep]) > 0)
+    # Check that MarkupSafe is returned as indirect dep of jinja2:
+    assert [
+        dep for dep in get_package_dependencies(
+            "python-for-android", recursive=True
+        )
+        if "MarkupSafe" in dep
+    ]
+    # Check that colorama is returned without the install condition when
+    # just getting the names:
+    assert "colorama" in get_dep_names_of_package("python-for-android")
+
+
+def test__get_system_python_executable():
+    # This is kind of absurd that we need this to run the test at all,
+    # but we can't test this inside tox's virtualenv:
+    pybin = _get_system_python_executable()
+
+    # The python binary needs to match our major version to be correct:
+    pyversion = subprocess.check_output([
+        pybin, "-c", "import sys; print(sys.version)"
+    ], stderr=subprocess.STDOUT)
+    try:
+        pyversion = pyversion.decode("utf-8", "replace")
+    except AttributeError:
+        pass
+    assert(pyversion.strip() == sys.version.strip())
+
+    # Helper function to run our function:
+    def run__get_system_python_executable(pybin):
+        # We want to see what _get_system_python_executable() does given
+        # a specific python, so we need to make it import it and run it:
+        cmd = [
+            pybin,
+            "-c",
+            "import importlib\n"
+            "import json\n"
+            "import os\n"
+            "import sys\n"
+            "sys.path = [os.path.dirname(sys.argv[1])] + sys.path\n"
+            "m = importlib.import_module(\n"
+            "    os.path.basename(sys.argv[1]).partition('.')[0]\n"
+            ")\n"
+            "print(m._get_system_python_executable())",
+            os.path.join(os.path.dirname(__file__), "..",
+                         "pythonforandroid", "pythonpackage.py"),
+        ]
+        # Actual call to python:
+        try:
+            result = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as e:
+            raise RuntimeError("call failed, with output: " + str(e.output))
```

Comment:
why masking the `CalledProcessError` exception with another one? I'm not sure what purpose it addresses?

Reply:
This code path is very likely to be hit with a very wide array of errors, so this just improves the backtrace standard output (which for a subprocess error by default won't print the failed process output, which usually is vital to understand why things went wrong at this particular location)

Reply:
My idea is to simplify the code and flow control in general. If a branch provide just little value, I'd rather kill it.
As often I like to apply the Pareto principle to coding or [YANGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).
So basically this code path seam to serve only the case where people are updating p4a code and break something, then they would somehow see the trace from this line and somehow it would help them more than seeing a `CalledProcessError`. So I'd say, let's drop it

Reply:
YANGNI doesn't apply when I needed it :smile: 

Seriously, I put this in because I hit this multiple times. The error output really is *extremely* useless with this. Are you **sure** you want this gone?

Reply:
"just little value" - I got the strong feeling you haven't actually checked the effect, because I'm not sure how that would apply. Let me show you:

**Default output:**
`E               subprocess.CalledProcessError: Command '['/usr/bin/python3.7', '-c', "import importlib\nimport json\nbla()\nimport os\nimport sys\nsys.path = [os.path.dirname(sys.argv[1])] + sys.path\nm = importlib.import_module(\n    os.path.basename(sys.argv[1]).partition('.')[0]\n)\nprint(m._get_system_python_executable())", '/home/jonas/Develop/python-for-android/tests/../pythonforandroid/pythonpackage.py']' returned non-zero exit status 1.`

*(seriously, can you tell anything about what went wrong from this?)*

**Altered output:**
```
>           raise RuntimeError("call failed, with output: " + str(e.output))`
E           RuntimeError: call failed, with output: Traceback (most recent call last):
E             File "<string>", line 3, in <module>
E           NameError: name 'bla' is not defined
```
*(gives traceback + exact line for the externally run code for the nested error that occured)*

Now if you play around with this test as I did when writing it and you **still** want me to remove this, I'll happily do so. Let me know what you think

Reply:
it's just that the more you wrap you test with some bloated helper method, the more you're not testing what you are supposed to test.
Let's leave it for now because it's a detail compared to the general test structure. Just didn't have a chance to properly wrap my head around it to suggest better way to address.
Just a pointer: have you played with unittest.mock in the past? Maybe it's time to take a look

Reply:
@AndreMiras hmmmm... yes I used mock, but this function gets the system python and this tests that it works in different  python environments. How would you mock this? I mean the point is to see if it works in different environments, how would you mock an entire environment with all intricate details that could interfere?

Open for other approaches to testing this, I just don't see any

Reply:
Summed up better, this is definitely an integration test. (and inherently needs to be I'm quite sure, it depends too much on the environment that I see any useful way of mocking it) Would it help if I added a comment that makes this more clear?

Reply:
Yes it makes sense that you cannot mock if you play with system python, subprocess :+1: I was imagining so but I was hoping not :stuck_out_tongue: 
Like I said I haven't started digging seriously into what you're doing yet. So I'm just challenging you and throwing ideas in the hope you suddenly find a genius idea to make the whole think much more simple without me having to wrap my head into it somehow :sweat_smile:

Reply:
Oh right that makes sense. I tend to somehow assume you already got a perfect improved verison laid out in your head and sometimes I'm a bit stumped then that I didn't see it yet :joy: lol. Keep on challenging :+1:

Reply:
for what it's worth, this is how it looks like now: https://github.com/kivy/python-for-android/blob/02c7780c15f35dd94eff2cf12c172c3308251ccf/tests/test_pythonpackage_basic.py#L176

I still have that indirect-raise in because I found the additional output useful, but this shouldn't be the roadblock (= I could also remove it) if we want to move things ahead

Let me know if you consider this conversation resolved, or still want the `raise` indirection out!